presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.tree;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterExpression
```

Comment:
This doesn't feel like the right approach. Why do we need a whole new expression type for this?

They syntax tree is meant for representing syntactic SQL structures, but this is purely internal planning/execution concept. My (long term) goal is to remove the remaining non-syntactic elements from the syntax tree and introduce a new IR for query plans, so this moves us in the wrong direction.

Reply:
What do you suggest instead? Initially I was thinking that dynamic filter could be just `FunctionCall` (e.g: `dynamic_filter(probeExpr, 'source_id', 'symbo_id')`), but I'm not sure if this will be simpler approach since there would be no such function in registry. Rather those function calls need to be captured in local execution table scan planning.

Reply:
1) I don't think that'd be unreasonable. We can always add an internal function that fails if it's ever executed.
2) Why does it need to be embedded in an expression? Couldn't it be modeled as a new PlanNode that produces a result that gets used in a Filter above it?

```
- Filter: $0 AND foo(...)
     - DynamicFilter: probeExpression, sourceId, symbolId -> $0
         - ...
```

Reply:
> Why does it need to be embedded in an expression? Couldn't it be modeled as a new PlanNode that produces a result that gets used in a Filter above it?

See https://github.com/prestodb/presto/pull/9453#issuecomment-351671085
and
https://github.com/prestodb/presto/pull/9453#issuecomment-352691646

Reply:
@martint The idea was to reuse existing predicate pushdown optimizer. Initially `DynamicFilter` was implemented as a function call.

Reply:
I do not understand this signature `boolean dynamic_filter(expression, 'sourceId', 'symbolId')` can you please explain? What is `expression` here, is this evaluated before passing value to `dynamic_filter`? What is `symbolId`? Why not to pass symbol directly?

Reply:
Maybe we could extend `FilterNode` to have another new field `dynamicFilter` which would be just of type `DynamicFilter` (no `Expression`, no `FunctionCall`)? That way we could extend predicate pushdown to support this and there would be no issue with AST or `FunctionRegistry`.

Reply:
Ok. Lets take the this simple query.

```
SELECT * 
FROM t1, t2
WHERE t1.c1 = t2.c2
```

We want to pre-filter `t1` rows based on the set of values from `t2.c2` 

We assign `t2` a name, called `dynamic_filter_source_table_2`. We assign `c2` a name as well, called `dynamic_filter_source_symbol_column_2`

Than we introduce dynamic filter

```
SELECT * 
FROM t1, t2
ON t1.c1 = t2.c2
WHERE dynamic_filter(t1.c1, 'dynamic_filter_source_table_2', 'dynamic_filter_source_symbol_column_2') = TRUE
```

assuming `dynamic_filter` function somehow magically knows how to get `c2` values from `t2` and does `contains`, those queries are equivalent. 

In this case `expression` you are asking is just a column reference to `t1.c1`. But it can be anything.

Now, we can run regular optimizer on that query. The problem is that we don't know how to get dynamic values from the build side using standard `function` capabilities. So further we replace this function call with a lot of execution magic. To make sure we do replace - this function can be stubbed with a `throw RuntimeException()`

Introducing another field to `FilterNode` will only partially solve the problem. You can have predicates in JoinNode as well (and maybe in some other nodes). 

Although function approach seems to be hacky - i think it is not that bad. I don't think it is worth changing a lot of code to just have it other way.

Reply:
> Initially DynamicFilter was implemented as a function call

@arhimondr do you mean `sym == deffered_symbol` or there was something else before?

> boolean dynamic_filter(expression, 'sourceId', 'symbolId')

That is fine (and seems cleaner) but I'm worried we are testing patience of the PR author this way ;)

> Maybe we could extend FilterNode to have another new field dynamicFilter which would be just of type ...

Then you need to treat such field specially on plan node level which we wanted to avoid.

Reply:
Yeah, i forgot how it was. It was `sym == deffered_symbol`, and `deffered_symbol` was expressed as a function call in serialization.

Reply:
yes. That sounds fine

Reply:
@arhimondr Will function call need any change in predicate pushdown ?
@martint Are you ok with this being a functional call ? Just trying to get everybody on same page here to prevent going back and forth.

Reply:
You might need to register a dummy stub in the FunctionRegistry

Reply:
@amoghmargoor, sorry for the delay. I didn’t see your question. Yes, a function call is ok.